pgx/src/array.rs

```diff
@@ -0,0 +1,215 @@
+            as _
+        }
+    }
+
+    /// A raw slice of the dimensions.
+    /// Oxidized form of ARR_DIMS(ArrayType*)
+    ///
+    /// # Safety
+    ///
+    /// This requires &mut to collect the slice, but be aware: this is a raw pointer!
+    /// If you find ways to store it, you are probably violating ownership.
+    /// Raw pointer validity is **asserted on dereference, not construction**,
+    /// so remember, this slice is only guaranteed to be valid almost immediately
+    /// after obtaining it, or if you continue to hold the RawArray.
+    pub fn dims(&mut self) -> NonNull<[libc::c_int]> {
+        // for expected behavior, see:
+        // postgres/src/include/utils/array.h
+        // #define ARR_DIMS
+
+        // SAFETY: Welcome to the infernal bowels of FFI.
+        // Because the initial ptr was NonNull, we can assume this is also NonNull.
+        // As validity of the initial ptr was asserted on construction of RawArray,
+        // this can assume the dims ptr is also valid, allowing making the slice.
+        // This code doesn't assert validity per se, but in practice,
+        // the caller will probably immediately turn this into a borrowed slice,
+        // opening up the methods that are available on borrowed slices.
+        //
+        // So, to be clear, yes, everything done so far allows the caller to do so,
+        // though it is possible the caller can misuse this in various ways.
+        // Only the "naive" case is well-guarded.
+        unsafe {
+            let len = self.ndims() as usize;
+            NonNull::new_unchecked(slice_from_raw_parts_mut(
+                pgx_ARR_DIMS(self.at.as_ptr()),
+                len,
+            ))
+        }
+    }
+
+    /// The flattened length of the array.
+    pub fn len(&self) -> libc::c_int {
+        // SAFETY: Validity asserted on construction, and...
+        // ...well, hopefully Postgres knows what it's doing.
+        unsafe {
+            pgx_ARR_NELEMS(self.at.as_ptr())
+            // FIXME: While this was indeed a function that returns a c_int,
+            // using a usize is more idiomatic in Rust, to say the least.
+            // In addition, the actual sizes are under various restrictions,
+            // so we probably can further constrain the value, honestly.
+            // However, PGX has trouble with returning usizes
+            as _
+        }
+    }
+
+    pub fn oid(&self) -> pg_sys::Oid {
+        // SAFETY: Validity asserted on construction.
+        unsafe { (*self.at.as_ptr()).elemtype }
+    }
+
+    /// Gets the offset to the ArrayType's data.
+    /// Note that this should not be "taken literally".
+    fn data_offset(&self) -> i32 {
+        // SAFETY: Validity asserted on construction.
+        unsafe { (*self.at.as_ptr()).dataoffset }
+        // This field is an "int32" in Postgres
+    }
+
+    /// Equivalent to ARR_HASNULL(ArrayType*)
+    /// Note this means that it only asserts that there MIGHT be a null
```

Comment:
`MIGHT` 🤔

Reply:
PERHAPS UNNECESSARY CAPITALIZATION, or?

Reply:
No, not at all, it's a good callout. It was more like "Im calling something that might have null" instead of "I am calling something I know for sure will be null or not".